HPA Controller: Use Custom Metrics API

DirectXMan12 opened the pull request:
This commit switches over the HPA controller to use the custom metrics
API.  It also converts the HPA controller to use the generated client
in k8s.io/metrics for the resource metrics API.

In order to enable support, you must enable
`--horizontal-pod-autoscaler-use-rest-clients` on the
controller-manager, which will switch the HPA controller's MetricsClient
implementation over to use the standard rest clients for both custom
metrics and resource metrics.  This requires that at the least resource
metrics API is registered with kube-aggregator, and that the controller
manager is pointed at kube-aggregator.  For this to work, Heapster
must be serving the new-style API server (`--api-server=true`).

Before this merges, this will need kubernetes/metrics#2 to merge, and a godeps update to pull that in.
It's also semi-dependent on kubernetes/heapster#1537, but that is not required in order for this to merge.

**Release note**:
```release-note
Allow the Horizontal Pod Autoscaler controller to talk to the metrics API and custom metrics API as standard APIs.
```

DirectXMan12 committed:
HPA Controller: Use Custom Metrics API

This commit switches over the HPA controller to use the custom metrics
API.  It also converts the HPA controller to use the generated client
in k8s.io/metrics for the resource metrics API.

In order to enable support, you must enable
`--horizontal-pod-autoscaler-use-rest-clients` on the
controller-manager, which will switch the HPA controller's MetricsClient
implementation over to use the standard rest clients for both custom
metrics and resource metrics.  This requires that at the least resource
metrics API is registered with kube-aggregator, and that the controller
manager is pointed at kube-aggregator.  For this to work, Heapster
must be serving the new-style API server (`--api-server=true`).

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41824)
<!-- Reviewable:end -->

DirectXMan12 commented:
@jszczepkowski PTAL
@kubernetes/rh-cluster-infra 

this is the last part of the HPA v2 changes that need to go in to k8s.io/kubernetes proper

derekwaynecarr commented on pkg/controller/podautoscaler/metrics/legacy_metrics_client.go:
```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2017

derekwaynecarr commented on pkg/controller/podautoscaler/metrics/legacy_metrics_client.go:
```diff
@@ -0,0 +1,212 @@
+
+var heapsterQueryStart = -5 * time.Minute
+
+type HeapsterMetricsClient struct {
```
is there a reason this needs to be public?

DirectXMan12 replied:
no, it was previously, though -- everything in `legacy` is just the previous client (literally `$ mv metrics_client.go legacy_metrics_client.go` and then move the interfaces off to their own place).

DirectXMan12 committed:
Add dependencency on k8s.io/metrics

This commit introduces a dependency on k8s.io/metrics, for
use with the horizontal pod autoscaler.

DirectXMan12 commented:
@k8s-bot verify test this

krzyzacy commented:
@k8s-bot gci gce e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot verify test this

that's a log of presubmits :-p

DirectXMan12 commented:
@k8s-bot gci gce e2e test this
@k8s-bot cvm gce e2e test this

flake #41889
flake #41898
flake #41902

DirectXMan12 commented:
verification needs `k8s.io/metrics` to resolve to something.  The repo is created, what needs to be done to get that to happen?

jszczepkowski commented:
General comment: it will be much cleaner if you leave "metrics_client.go" as it is, and introduce a new client in file with a different name, e.g.: "rest_metrics_client.go"

DirectXMan12 commented:
Ah, sadly, GitHub appears to have not recognized the rename -- that's effectively what I did -- I renamed metrics_client.go to legacy_metrics_client.go

DirectXMan12 commented:
alright, fixed the naming.  GitHub should show the renames properly now.

jszczepkowski commented:
Metrics client unit test should be extended to test the new client.

DirectXMan12 commented:
unit test flake appears to be #41961
Not sure what's up with the verification

DirectXMan12 commented:
so, I think what's going on with the godeps is that Godep doesn't like that `k8s.io/metrics` is relying on things from the staging directory, which, by virtue of being in the staging directory, aren't in the Godeps file.  I'm hitting the same symptom as @lavalamp in #41269, but I'm not convinced it's the same issue.

cc @kubernetes/sig-api-machinery-misc any ideas on what to do here?

DirectXMan12 commented:
flake seen in #37891 (possibly is #35771)
I'll hold off rerunning until we get the verification sorted out

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,136 @@
+			if !found {
+				missing = true
+				glog.V(2).Infof("missing resource metric %v for container %s in pod %s/%s", resource, c.Name, namespace, m.Name)
+				continue
```
s/continue/break
If missing is true, all computations will be discarded

jszczepkowski replied:
You can use break with a label to clarify which for will be broken.

DirectXMan12 replied:
ack.  That's the same way in the legacy client.

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client_test.go:
```diff
@@ -0,0 +1,274 @@
+	desiredMetricValues PodMetricsInfo
+	desiredError        error
+
+	targetTimestamp      int
```
Please add a comment explaining unit of this variable

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client_test.go:
```diff
@@ -0,0 +1,274 @@
+
+	assert.Equal(t, tc.desiredMetricValues, metrics, "the metrics values should be as expected")
+
+	targetTimestamp := fixedTimestamp.Add(time.Duration(tc.targetTimestamp) * time.Minute)
```
I don't understand, why Duration is used here. It seems to be redundant and wrong.

DirectXMan12 replied:
you have to cast to a `time.Duration` before you can multiply it by a `time.Duration` (which `time.Minute` is).  It's the same way in the legacy rest client test.

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client_test.go:
```diff
@@ -0,0 +1,274 @@
+						Namespace: namespace,
+						Labels:    podLabels,
+					},
+					Timestamp:  metav1.Time{Time: fixedTimestamp.Add(time.Duration(tc.targetTimestamp) * time.Minute)},
```
I don't understand, why Duration is used here. It seems to be redundant and wrong.

DirectXMan12 replied:
(see above).

jszczepkowski commented:
One general comment: unit-tests horizontal_test.go and replica_calculator_test.go were modified to test only the case of rest metrics client, leaving legacy client not tested. I think it is wrong, we should leave legacy tests present as long as legacy client exists. We can possible change their names to legacy_horizontal_test.go and legacy_replica_calculator_test.go.

jszczepkowski commented:
I've completed the first pass, however, I'm not expert on APIs/clients. We should find someone to review and approve the second commit. Maybe @deads2k or @smarterclayton?

DirectXMan12 commented:
> We can possible change their names to legacy_horizontal_test.go and legacy_replica_calculator_test.go.

Will do.

DirectXMan12 committed:
Add staging repos to GOPATH in verify-godeps

This commit adds the staging repos to the GOPATH in
hack/verify-godeps.sh.  This allows vendored librarys
to depend on staging repos and not break verify-godeps.

This also adds `hack/godep-restore.sh`, which acts like
`godep restore`, but sets the GOPATH appropriate as well.

DirectXMan12 commented:
cc @sttts @deads2k @ncdc here's the new commit with the `hack/verify-godeps.sh` change

DirectXMan12 commented:
*grumble grumble* there's a script different from `verify-godeps` that runs `godep restore` before running `verify-godeps` *grumble grumble*

DirectXMan12 commented:
@jsz any other comments?
@kubernetes/sig-api-machinery-pr-reviews can one of you do a sanity check on this
@deads2k @mtaufen PTAL or point me in the direction of someone else to bug for approval

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,136 @@
+	client resourceclient.PodMetricsesGetter
+}
+
+func (c *resourceMetricsClient) GetResourceMetric(resource v1.ResourceName, namespace string, selector labels.Selector) (PodMetricsInfo, time.Time, error) {
```
Please add comment

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,136 @@
+	client customclient.CustomMetricsClient
+}
+
+func (c *customMetricsClient) GetRawMetric(metricName string, namespace string, selector labels.Selector) (PodMetricsInfo, time.Time, error) {
```
Please add comment

jszczepkowski commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,136 @@
+	return res, timestamp, nil
+}
+
+func (c *customMetricsClient) GetObjectMetric(metricName string, namespace string, objectRef *autoscaling.CrossVersionObjectReference) (int64, time.Time, error) {
```
Please add comment

jszczepkowski commented:
I have only some nits. After applying them: LGTM for HPA part.

jszczepkowski commented:
One more comment: please mark in git as a file rename:
* horizontal_test.go -> leagcy_horizontal_test.go and
* replica_calculator_test.go -> legacy_replica_calculator_test.go
They are currently treated as new files.

jszczepkowski commented:
E2E tests for HPA should be extended, to run in two configurations: legacy and rest. Currently only legacy code is tested by e2e tests. (This, of course, should be done in a subsequent PR).

derekwaynecarr commented on pkg/apis/componentconfig/types.go:
```diff
@@ -796,6 +796,10 @@ type KubeControllerManagerConfiguration struct {
 	EnableTaintManager bool
 	// If set to true NodeController will use taints to evict Pods from notReady and unreachable Nodes.
 	UseTaintBasedEvictions bool
+	// horizontalPodAutoscalerUseRESTClients causes the HPA controller to use REST clients
```
nit: s/horizontal/Horizontal

derekwaynecarr replied:
i actually find this description clearer than what you use to document the flag in options.go

derekwaynecarr commented on pkg/controller/podautoscaler/metrics/interfaces.go:
```diff
@@ -0,0 +1,45 @@
+
+// MetricsClient knows how to query a remote interface to retrieve container-level
+// resource metrics as well as pod-level arbitrary metrics
+type MetricsClient interface {
```
i am still parsing through, and this could be handled in any later refactor, but is there a reason we didnt have:

`client.Metrics(namespace).GetResourceMetric(resource, selector)`

as the API pattern?

DirectXMan12 replied:
There's no particular reason, just that I didn't want to add a new implementation and refactor the existing interface at the same time.

DirectXMan12 replied:
I'll post a follow-on PR at a later date

derekwaynecarr commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,142 @@
+	}
+}
+
+// FullMetricsClient is a client which supports fetching
```
what is a FullMetricsClient?  do you mean a MetricsClient or a restMetricsClient in this context?

DirectXMan12 replied:
I meant restMetricsClient -- I previously called it FullMetricsClient, decided against that name, and forgot to update the godoc ;-)

derekwaynecarr commented on pkg/controller/podautoscaler/metrics/rest_metrics_client.go:
```diff
@@ -0,0 +1,142 @@
+// GetResourceMetric gets the given resource metric (and an associated oldest timestamp)
+// for all pods matching the specified selector in the given namespace
+func (c *resourceMetricsClient) GetResourceMetric(resource v1.ResourceName, namespace string, selector labels.Selector) (PodMetricsInfo, time.Time, error) {
+	metrics, err := c.client.PodMetricses(namespace).List(metav1.ListOptions{LabelSelector: selector.String()})
```
per our chat on irc, this sounds like gollum "metricses"...

are we stuck with this name forever if its in v1alpha1 clients already?

it would be good if we can get a fix post feature freeze to have it sound like sensible english.

DirectXMan12 replied:
I'll post a follow up PR -- it needs changes to gengo to detect already-plural type names and not attempt to pluralize them.

smarterclayton replied:
I'm ok with a fix going in post code freeze

derekwaynecarr reviewed (commented):
generally looks good, i would like us to fix the client name generation to not sounds like gollum before release post freeze.

smarterclayton commented:
/approve

Based on comments from various responsible teams and reviewers

smarterclayton commented:
It looked like all comments were addressed - anyone else have any last minute things they need here?

derekwaynecarr commented:
@smarterclayton -- this is lgtm for me.

derekwaynecarr commented:
tagging per comment thread.

/lgtm

derekwaynecarr commented:
/lgtm

derekwaynecarr commented:
/lgtm

derekwaynecarr commented:
/lgtm

ncdc commented:
Re-adding lgtm post rebase

sttts commented:
This breaks godep-restore: https://github.com/kubernetes/kubernetes/issues/42403

DirectXMan12 commented:
You need to use the `hack/godep-restore.sh` script

sttts commented on Godeps/Godeps.json:
```diff
@@ -2791,6 +2791,58 @@
 			"Rev": "c2ac40f1adf8c42a79badddb2a2acd673cae3bcb"
 		},
 		{
+			"ImportPath": "k8s.io/metrics/pkg/apis/custom_metrics",
```
Adding k8s.io/metrics as a vendored package was not a good idea, leading to circular repo dependencies: it depends on k8s.io/apimachinery which is in staging/ here. Changes in apimachinery easily trigger compile errors in vendor/k8s.io/metrics. It's more than tedious to fix them by updating k8s.io/metrics first with a new apimachinery and then re-vendor it here.

sttts replied:
@DirectXMan12 I have to generate deepcopies in k8s.io/metrics, but do not see any scripts for that.

DirectXMan12 replied:
Should be in vendor/k8s.io/repo-infra/generate/

DirectXMan12 replied:
> Adding k8s.io/metrics as a vendored package was not a good idea, leading to circular repo dependencies: it depends on k8s.io/apimachinery which is in staging/ her

Well, it was desired not to have the type definitions be part of the main repository, but we need to consume apimachinery, etc as part of the API machinery for registering the API types, having autogenerated clients, etc, so I'm not sure what the solution is there (unless we want to make k8s.io/metrics a staging repo...).

caesarxuchao replied:
I'm hitting this circular dependency. kubernetes->metric->client-go (which is in kubernetes/staging).

I'm going to generate the metrics client to untangle part of the dependency.

There are other circular dependencies caused by vendoring k8s.io/metrics. IMO we should push to remove the staging folder and develop in the real repos to break the cycle.

cc @thockin another problem caused by staging.

caesarxuchao replied:
The entanglement is more complex than i thought. Apart from the core client, k8s.io/metrics is [importing]("https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go#L21") "k8s.io/client-go/pkg/api/v1" as well. It's going to block us from moving external types to another repo in https://github.com/kubernetes/kubernetes/pull/44784.

One expedient solution is moving the entire vendor/k8s.io/metric to thirdparty/forked and manually take care the dependency until we get rid of the staging. @sttts @DirectXMan12 @lavalamp how's that sound?

caesarxuchao replied:
Sent #45159. Still WIP.

DirectXMan12 replied:
I think I'd probably be in favor of moving k8s.io/metrics to staging.  In the short term, it'll also make regenerating the types easier (although I'll need to port a couple of my generator changes, which add pretty simple new tags).

caesarxuchao replied:
What do you mean by "regenerating the types"? May i ask why it is difficult to regenerate if we move it to thirdparty?

Staging uses the simlink trick to appear as vendored. I prefer not using staging if a plain directory like thirdparty/ works.

BTW, do you know if the clientset is generated in k8s.io/metrics? I tried to run client-gen and the output is quite different.

DirectXMan12 replied:
"regenerating the types" == generating clientset, conversions, protobuf, etc.

`k8s.io/metrics/pkg/apis/custom_metrics` uses a hendwritten client, `k8s.io/metrics/pkg/apis/metrics` uses an autogenerated client.  Both use autogenerated conversions, etc.  The copy of clientgen used to generate clients for `pkg/apis/metrics` was extracted from the main repo a while ago, and then has a couple of improvements tacked on to handle read-only APIs and such.  Unfortunately, it currently only exists in the form of https://github.com/kubernetes/repo-infra/pull/7, because I haven't had a chance to fix up the rest of the stuff in that PR.

I figured if we moved to staging, I could just port the two tweaks to client-gen over to the main repo's client-gen, and then just use the main repo's code gen scripts to generate conversions, etc (and then sync stuff back to the main k8s.io/metrics repo as per the staging sync process).

In the long run, we'll probably need to make out-of-main-repo codegen easier, but that work will probably come as part of the new https://github.com/kubernetes-incubator/apiserver-builder.

caesarxuchao replied:
> and then sync stuff back to the main k8s.io/metrics repo as per the staging sync process

this part is not easy. I'll need to reconfigure the publishing robot and i don't know if it will work as expected :)
